src/systems/elevator/utils/DoorTimer.hh

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (C) 2021 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+/*
+ * \author Nick Lamprianidis <nlamprian@gmail.com>
+ * \date January 2021
+ */
+
+#ifndef IGNITION_GAZEBO_SYSTEMS_DOOR_TIMER_HH_
+#define IGNITION_GAZEBO_SYSTEMS_DOOR_TIMER_HH_
+
+#include <chrono>
+#include <functional>
+#include <memory>
+
+#include <ignition/gazebo/System.hh>
+
+namespace ignition
+{
+namespace gazebo
+{
+// Inline bracket to help doxygen filtering
+inline namespace IGNITION_GAZEBO_VERSION_NAMESPACE
+{
+namespace systems
+{
+// Data forward declaration
+class DoorTimerPrivate;
+
+/// \brief Timer that's used to keep a door open. It has a configurable default
+/// wait duration that when exceeded, it calls a function to let the state
+/// machine know to transition to the next state. The timer also checks whether
+/// the doorway is blocked, in which case it keeps the door open until whatever
+/// blocks the doorway moves out of the way.
+class DoorTimer
+{
+  /// \brief Constructor
+  public: DoorTimer();
+
+  /// \brief Destructor
+  public: ~DoorTimer();
+
+  /// \brief Sets the default wait duration
```

Comment:
It's good to document that `Configure` needs to be called after this, otherwise the timeout won't be updated.

A higher level recommendation though would be to get rid of the `SetWaitDuration` and `Configure` functions and instead do it all in the constructor to follow the RAII idiom. The duration can have a default value so it's optional. This way, there's no way the user can call functions in the wrong order or forget to call them.

Reply:
I agree in regards to SetWaitDuration. I did it like this because the duration doesn't exist at the moment of creation. Not sure if I can consider it optional and give it a default value, so I think I'll turn the timer variable into a unique_ptr to get around that.

For Configure though, it's called every time the timer is started, so it needs to stay. Does it make you draw a correlation to ISystemConfigure by giving it this name? The same concept applies to the joint monitor.